In [6]:
!git clone https://github.com/yuanzhumia0/CA1-2026025-Machine-Learning-for-Business.git

Cloning into 'CA1-2026025-Machine-Learning-for-Business'...


In [2]:
%pip install pandas pyarrow

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd

df = pd.read_parquet(
    "/Users/cat/Documents/GitHub/CA1-2026025-Machine-Learning-for-Business/yellow_tripdata_2025-01.parquet"
)


Check how many rows 

In [5]:
df.head() 

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,1,2025-01-01 00:18:38,2025-01-01 00:26:59,1.0,1.60,1.0,N,229,237,1,10.0,3.5,0.5,3.00,0.0,1.0,18.00,2.5,0.0,0.0
1,1,2025-01-01 00:32:40,2025-01-01 00:35:13,1.0,0.50,1.0,N,236,237,1,5.1,3.5,0.5,2.02,0.0,1.0,12.12,2.5,0.0,0.0
2,1,2025-01-01 00:44:04,2025-01-01 00:46:01,1.0,0.60,1.0,N,141,141,1,5.1,3.5,0.5,2.00,0.0,1.0,12.10,2.5,0.0,0.0
3,2,2025-01-01 00:14:27,2025-01-01 00:20:01,3.0,0.52,1.0,N,244,244,2,7.2,1.0,0.5,0.00,0.0,1.0,9.70,0.0,0.0,0.0
4,2,2025-01-01 00:21:34,2025-01-01 00:25:06,3.0,0.66,1.0,N,244,116,2,5.8,1.0,0.5,0.00,0.0,1.0,8.30,0.0,0.0,0.0


Check the number of columns.

In [7]:
df.shape[1]

20

Get all column names of the DataFrame.

In [8]:
df.columns.tolist()

['VendorID',
 'tpep_pickup_datetime',
 'tpep_dropoff_datetime',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'store_and_fwd_flag',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'total_amount',
 'congestion_surcharge',
 'Airport_fee',
 'cbd_congestion_fee']

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3475226 entries, 0 to 3475225
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

### January Data Overview

The January dataset has 3,475,226 rows and 20 columns. Each row contains information about one taxi trip, such as the trip distance, pickup time, drop-off time, and fare.

The pickup and drop-off times are already in datetime format. I can use them later to calculate how long each trip lasted.

Next, I will check for missing values.

### Check missing values in each column

In [10]:
missing_values = pd.DataFrame({
    "Missing values": df.isna().sum(),
    "Missing (%)": (df.isna().mean() * 100).round(2)
})

missing_values

,Missing values,Missing (%)
VendorID,0,0.00
tpep_pickup_datetime,0,0.00
tpep_dropoff_datetime,0,0.00
passenger_count,540149,15.54
trip_distance,0,0.00
RatecodeID,540149,15.54
store_and_fwd_flag,540149,15.54
PULocationID,0,0.00
DOLocationID,0,0.00
payment_type,0,0.00


### Missing Values

The result shows that most columns do not have missing values.

However, `passenger_count`, `RatecodeID`, `store_and_fwd_flag`, `congestion_surcharge`, and `Airport_fee` have missing values.

Each of these columns has **540,149 missing values**, which is **15.54%** of the total data.

These missing values may need to be handled before further analysis.

In [11]:
# Summary statistics for numerical columns
df.describe().T

,count,mean,min,25%,50%,75%,max,std
VendorID,3475226.0,1.785428,1.0,2.0,2.0,2.0,7.0,0.426328
tpep_pickup_datetime,3475226,2025-01-17 11:02:55.910964,2024-12-31 20:47:55,2025-01-10 07:59:01,2025-01-17 15:41:33,2025-01-24 19:34:06,2025-02-01 00:00:44,NaN
tpep_dropoff_datetime,3475226,2025-01-17 11:17:56.997901,2024-12-18 07:52:40,2025-01-10 08:15:29.500000,2025-01-17 15:59:34,2025-01-24 19:48:31,2025-02-01 23:44:11,NaN
passenger_count,2935077.0,1.297859,0.0,1.0,1.0,1.0,9.0,0.75075
trip_distance,3475226.0,5.855126,0.0,0.98,1.67,3.1,276423.57,564.6016
RatecodeID,2935077.0,2.482535,1.0,1.0,1.0,1.0,99.0,11.632772
PULocationID,3475226.0,165.191576,1.0,132.0,162.0,234.0,265.0,64.529483
DOLocationID,3475226.0,164.125177,1.0,113.0,162.0,234.0,265.0,69.401686
payment_type,3475226.0,1.036623,0.0,1.0,1.0,1.0,5.0,0.701333
fare_amount,3475226.0,17.081803,-900.0,8.6,12.11,19.5,863372.12,463.472918


### Summary Statistics

The summary statistics show the main features of the numerical data.

The average passenger count is about **1.30**, and the average trip distance is about **5.86 miles**.

The average fare amount is about **17.08**, while the average total amount is about **25.61**.

Some very large, small, and negative values can also be seen in the data. These may be outliers and should be checked before further analysis.

In [12]:
# Check for duplicate rows
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


### Duplicate Records¶

I checked for rows with identical values in all columns. No duplicate rows were found, so no rows were removed in this step.

In [13]:
# Keep trips that started in January 2025
df_clean = df[
    (df["tpep_pickup_datetime"] >= "2025-01-01") &
    (df["tpep_pickup_datetime"] < "2025-02-01")
].copy()

print("Rows before filtering:", len(df))
print("Rows after filtering:", len(df_clean))
print("Rows removed:", len(df) - len(df_clean))

Rows before filtering: 3475226
Rows after filtering: 3475204
Rows removed: 22


### Selecting January Trips

I kept trips with pickup times in January 2025. This removed 22 records and left 3,475,204 trips for further analysis.

In [14]:
# Calculate trip duration in minutes
df_clean["trip_duration_minutes"] = (
    df_clean["tpep_dropoff_datetime"] -
    df_clean["tpep_pickup_datetime"]
).dt.total_seconds() / 60

# Display summary statistics
print(df_clean["trip_duration_minutes"].describe())

# Count trips with zero or negative duration
print(
    "\nTrips with duration <= 0:",
    (df_clean["trip_duration_minutes"] <= 0).sum()
)

count    3.475204e+06
mean     1.501811e+01
std      3.871367e+01
min     -5.147232e+04
25%      7.283333e+00
50%      1.170000e+01
75%      1.833333e+01
max      5.626317e+03
Name: trip_duration_minutes, dtype: float64

Trips with duration <= 0: 2051


### Trip Duration

I calculated the trip duration in minutes using the pickup and drop-off times.

The average trip duration is about **15.02 minutes**, and the median duration is about **11.7 minutes**.

There are **2,051 trips** with zero or negative duration. These records are not valid, so they will be removed during data cleaning.

There are also some trips with very long durations, which may be outliers and need further checking.

In [15]:
# Remove trips with zero or negative duration
rows_before = len(df_clean)

df_clean = df_clean[
    df_clean["trip_duration_minutes"] > 0
].copy()

print("Rows removed:", rows_before - len(df_clean))
print("Rows remaining:", len(df_clean))

Rows removed: 2051
Rows remaining: 3473153


In [16]:
# Check the upper percentiles of trip duration
df_clean["trip_duration_minutes"].quantile([0.95, 0.99, 0.999])

0.950    36.216667
0.990    58.916667
0.999    98.280800
Name: trip_duration_minutes, dtype: float64

### Checking Long Trips

I removed 2,051 trips with zero or negative duration, leaving 3,473,153 records.

About 99% of trips lasted no more than 58.92 minutes, and 99.9% lasted no more than 98.28 minutes. Longer trips may be unusual, but they are not necessarily errors. I will check the longest trips before deciding whether to remove them.

In [17]:
# Inspect the 10 longest trips
columns_to_check = [
    "tpep_pickup_datetime",
    "tpep_dropoff_datetime",
    "trip_duration_minutes",
    "trip_distance",
    "fare_amount",
    "total_amount"
]

df_clean.nlargest(10, "trip_duration_minutes")[columns_to_check]

,tpep_pickup_datetime,tpep_dropoff_datetime,trip_duration_minutes,trip_distance,fare_amount,total_amount
1858816,2025-01-21 12:23:41,2025-01-25 10:10:00,5626.316667,6.55,-28.20,-36.45
1858817,2025-01-21 12:23:41,2025-01-25 10:10:00,5626.316667,6.55,28.20,36.45
414374,2025-01-06 02:12:11,2025-01-09 13:54:19,5022.133333,91.51,336.90,339.40
2460536,2025-01-27 11:58:55,2025-01-30 09:42:55,4184.000000,0.00,3.00,4.50
898201,2025-01-11 08:04:00,2025-01-13 16:49:28,3405.466667,0.04,70.00,76.50
2147536,2025-01-24 07:07:21,2025-01-26 10:09:44,3062.383333,0.00,3.00,4.50
85681,2025-01-02 09:40:31,2025-01-04 09:37:01,2876.500000,0.00,3.00,4.50
1536147,2025-01-17 16:45:49,2025-01-19 16:40:59,2875.166667,1.98,12.80,20.05
556772,2025-01-07 18:07:33,2025-01-09 16:12:05,2764.533333,1.86,47.80,51.80
3417048,2025-01-29 21:39:23,2025-01-31 16:48:43,2589.333333,7.52,31.02,35.77


### Removing Extremely Long Trips

Some of the longest records lasted several days but had short or zero trip distances. These records appear unsuitable for analysing typical taxi trips.

For this project, I will keep trips lasting up to 180 minutes. This is a practical limit for my analysis, although it may exclude some valid long trips.

In [18]:
# Keep trips lasting up to 3 hours
rows_before = len(df_clean)

df_clean = df_clean[
    df_clean["trip_duration_minutes"] <= 180
].copy()

print("Rows removed:", rows_before - len(df_clean))
print("Rows remaining:", len(df_clean))

Rows removed: 1377
Rows remaining: 3471776


After applying the 180-minute limit, I removed 1,377 records. The dataset now contains 3,471,776 trips, each with a duration greater than zero and no more than 180 minutes.

In [19]:
# Keep trips starting in January
df_clean = df[
    (df["tpep_pickup_datetime"] >= "2025-01-01") &
    (df["tpep_pickup_datetime"] < "2025-02-01")
].copy()

# Calculate trip duration in minutes
df_clean["trip_duration_minutes"] = (
    df_clean["tpep_dropoff_datetime"] -
    df_clean["tpep_pickup_datetime"]
).dt.total_seconds() / 60

# Keep durations greater than 0 and no more than 180 minutes
df_clean = df_clean[
    (df_clean["trip_duration_minutes"] > 0) &
    (df_clean["trip_duration_minutes"] <= 180)
].copy()

print("Rows remaining:", len(df_clean))

Rows remaining: 3471776


### Cleaning Trip Duration

I kept trips that started in January 2025 and calculated their duration in minutes.

I removed 2,051 trips with zero or negative duration and 1,377 trips lasting more than 180 minutes. After these steps, 3,471,776 records remain.

Next, I will check trip distances and fares for unusual values.

In [20]:
# Check non-positive distances and fares
print("Trips with distance <= 0:",
      (df_clean["trip_distance"] <= 0).sum())

print("Trips with fare <= 0:",
      (df_clean["fare_amount"] <= 0).sum())

print("Trips with total amount <= 0:",
      (df_clean["total_amount"] <= 0).sum())

# Check percentiles and maximum values
df_clean[["trip_distance", "fare_amount", "total_amount"]].quantile(
    [0.50, 0.95, 0.99, 0.999, 1.0]
)

Trips with distance <= 0: 90172
Trips with fare <= 0: 145441
Trips with total amount <= 0: 63535


,trip_distance,fare_amount,total_amount
0.500,1.67,12.11,19.95000
0.950,11.83,52.00,74.00000
0.990,19.50,72.30,102.92000
0.999,28.82,131.10,163.24225
1.000,276423.57,863372.12,863380.37000
